src/ucp/rndv/proto_rndv.c

```diff
@@ -585,18 +591,23 @@ void ucp_proto_rndv_receive_start(ucp_worker_h worker, ucp_request_t *recv_req,
         ucp_proto_rndv_check_rkey_length(rts->address, rkey_length, "rts");
         op_id            = UCP_OP_ID_RNDV_RECV;
         recv_req->status = UCS_OK;
-        UCS_PROFILE_CALL_VOID(ucp_datatype_iter_init_from_dt_state,
-                              worker->context, recv_req->recv.buffer, rts->size,
-                              recv_req->recv.datatype, &recv_req->recv.state,
-                              &req->send.state.dt_iter, &sg_count);
+        size             = rts->size;
     } else {
         /* Short receive: complete with error, and send reply to sender */
         rkey_length      = 0; /* Override rkey length to disable data fetch */
         op_id            = UCP_OP_ID_RNDV_RECV_DROP;
         recv_req->status = UCS_ERR_MESSAGE_TRUNCATED;
+        size             = 0;
+    }
+
+    if (ucs_likely(size != 0)) {
```

Comment:
what is it fixing? some leak with rndv/ats? if yes, it's probably not the right way to fix since this code cannot make any assumptions on the protocol that will be selected

Reply:
what is the problem with initializing iterator when RTS has no data? it should still work, just like user can send 0-length buffer in any operation.

Reply:
the particular problem was that we do not call `dt_generic.ops.finish` if zero sized RTS arrives. I think it is more reasonable to  avoid initialization in such case, rather than doing init/finalize

Reply:
rndv/ats protocol should call finish, just like other protocols for RNDV_RECV opcode call finish.

Reply:
ok, then the same needs to be applied to RECV_DROP opcode and dt cleanup should be removed from this func completely